spec/datadog/appsec/rate_limiter_spec.rb

```diff
@@ -0,0 +1,38 @@
+require 'datadog/appsec/spec_helper'
+require 'datadog/appsec/rate_limiter'
+
+RSpec.describe Datadog::AppSec::RateLimiter do
+  before { described_class.reset!(:traces) }
+
+  describe '#limit' do
+    context 'in different threads' do
+      before do
+        allow(described_class).to receive(:trace_rate_limit).and_return(rate_limit)
+      end
+
+      let(:rate_limit) { 50 }
+      let(:spy_1) { spy('thread-1 workload') }
+      let(:spy_2) { spy('thread-2 workload') }
+
+      it 'creates separate rate limiter per thread' do
+        thread_1 = Thread.new do
+          Array.new(rate_limit * 2) do
+            described_class.limit(:traces) { spy_1.call }
+          end
+        end
```

Comment:
Is it possible that thread_1 executes completely before thread_2 starts, given that MRI is not necessarily concurrent fairly? What do you think about adding some `sleep 0` in the workload?

Reply:
Yes, it is, but the thing is even so, we will see limiting in second thread if they share the limiter. And the point is to show the split of limiter per thread.

Reply:
So I was summoned to this thread by seeing the idea of `sleep` in specs :rofl: 

`sleep`s tend to cause a lot of flaky specs in our test suite.

In the specific case we're discussing of `sleep 0`, I think you may want `Thread.pass` instead, since it's a stronger hint to the VM to "let other things run".

But to be honest, I think instead of this, you may want to go a bit more "white-box" testing, e.g. instead of testing that two threads can both use the rate limit, consider instead:
* Checking that two rate limit objects get created + (one per thread)
* Checking that each thread only plays with its own rate limit object

There's multiple ways of doing this, happy to help if it's not something you've done before :)

Reply:
`Thread.pass` is what I should have said, I keep forgetting it exists.

Reply:
@ivoanjo In 99.9% of the cases I'm against `sleep` in tests. And this test do exactly like you've described, Threads are using `spy` object and run more than allowed calls (x4 time in total) and I'm expecting both `spy` objects to receive similar amount of calls.

Because if they share the limiter, one of them will get less or non calls allowed.

Reply:
Gotcha, let me try to do so.

Reply:
Fixed https://github.com/DataDog/dd-trace-rb/pull/3975/commits/6ef9fd2ee66c4cdcf688c0a626f7bc45e986b9ba